sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/AwsPojoRow.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.values;
+
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+import static org.apache.beam.sdk.io.aws2.schemas.AwsPojoSchema.CACHE_IDX_OPTION;
+import static org.apache.beam.sdk.io.aws2.schemas.AwsPojoSchema.CACHE_SIZE_OPTION;
+import static software.amazon.awssdk.core.protocol.MarshallingType.LIST;
+import static software.amazon.awssdk.core.protocol.MarshallingType.MAP;
+import static software.amazon.awssdk.core.protocol.MarshallingType.SDK_POJO;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.function.Function;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.Schema.Field;
+import org.apache.beam.sdk.schemas.Schema.FieldType;
+import org.apache.commons.lang3.ArrayUtils;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import software.amazon.awssdk.core.SdkField;
+import software.amazon.awssdk.core.SdkPojo;
+import software.amazon.awssdk.core.protocol.MarshallingType;
+import software.amazon.awssdk.core.traits.ListTrait;
+import software.amazon.awssdk.core.traits.MapTrait;
+
+/**
+ * A row delegating to an underlying AWS {@link SdkPojo}.
+ *
+ * <p>This is following the example of {@link RowWithGetters}, but optimized for usage with {@link
+ * SdkPojo}s.
```

Comment:
Could you summarize what are the gaps in `RowWithGetters` that you need to address with this implementation? I'm wondering if these are fixes we should apply in  `RowWithGetters` instead of creating an alternate version.

It's worth taking the time to enumerate them - if we really do need to fork `RowWithGetters` in this way, I think we should document what is different and why in this javadoc.

Reply:
Thanks so much for looking into this @TheNeuralBit 🙏 

For completeness, I wasn't able to use the existing `JavaFieldSchema` or `JavaBeanSchema` for the following reasons:
- Model fields of the AWS SDK are private (and final).
- Getters / Setters lack expected prefixes to be valid beans.
- (Immutable) instances must be created using a builder (but these obviously lack the expected annotations).
- Some SDK models are self recursive, these are a major annoyance and required logic to eventually skip nested recursive fields as Beam schemas don't support this.

My initial version was implemented using `RowWithGetters`, but it required a couple of work arounds and seemed sub-optimal:

- I faced a mismatch when using logical types and I'm more and more wondering if that's a bug: `RowWithGetters` expects the getter to return the base type, but the target object will contain the input type (unless I misunderstand) forcing a pointless conversion input type -> base type -> input type ([see here](https://github.com/apache/beam/pull/11074#discussion_r818731944)). Particularly when working with collection types that seemed very inefficient.

- The getter factory `Factory<List<FieldValueGetter>>` doesn't work well with AWS model classes. Model instances already contain the necessary metadata (`List<SdkField<?>> sdkFields()`) including getters, making traversal of [nested models / rows](https://github.com/mosche/beam/blob/BEAM-13416-AwsSchemas/sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/AwsPojoRow.java#L117-L123) trivial. Using the getter factory, that field metadata isn't easily accessible as [factories operate on the class & schema](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java#L66). 

Additionally, there's a couple of optimisations. `RowWithGetters` might also benefit from these:

- [Nested rows](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java#L111) should be cached, otherwise the cache of such nested rows can't be leveraged

- Handling of collections / maps / iterables can be significantly improved in all cases where `getValue` is the [identity transform](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java#L142). In such cases the original collection / map / iterable can be returned as is (also making caching obsolete in that case).

- Caching using [3 separate hash maps](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java#L57-L59) comes with very high overhead for small rows (3 * `initialCapacity`). I figured I could use a single `Object[]` for caching as it's possible to [precalculate the required cache size and indices](https://github.com/mosche/beam/blob/BEAM-13416-AwsSchemas/sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/schemas/AwsPojoSchema.java#L242-L245) and store this using schema & field options.

Reply:
Thank you this is great!

I think it's fine to build custom SchemaProviders (rather than `JavaFieldSchema` and `JavaBeanSchema`)  for the AWS types. It's a common accepted pattern to build SchemaProviders for different domains (e.g. the protobuf type SchemaProvider in that protobuf extensions). It would be nice if an existing one "just worked", but the bar for creating a new one should be low IMO, as long as we take care to re-use things when appropriate.

> Some SDK models are self recursive, these are a major annoyance and required logic to eventually skip nested recursive fields as Beam schemas don't support this.

Is it OK to just skip those recursive fields?

> I faced a mismatch when using logical types...

Hm yeah this is a frustrating issue that's come up in several other contexts. Let's continue discussing it on the [source PR](https://github.com/apache/beam/pull/11074#discussion_r818731944). It would be nice to find a solution that's generally applicable.

> The getter factory `Factory<List<FieldValueGetter>>` doesn't work well with AWS model classes. Model instances already contain the necessary metadata ...

I'm not quite clear on this one, I guess since I have limited context on AWS model classes vs. Model instances. Is there some reference you could link for this?

> Additionally, there's a couple of optimisations. RowWithGetters might also benefit from these:

Could we separate out these changes and apply them to `RowWithGetters` in a way that's re-usable here?

Reply:
@TheNeuralBit As suggested I ported some of the changes to `RowWithGetters` https://github.com/apache/beam/pull/17172

Reply:
Regarding the getter factory, my issue is that it operates on the class level:
```java
getters = fieldValueGetterFactory.create(getterTarget.getClass(), schema);
```

Assuming the hypothetical AWS model class `Message` below, I need a getter factory that operates on the instance level to read `sdkFields()`:

```java
getters = fieldValueGetterFactory.create(getterTarget, schema);
```

Otherwise I have to read that per reflection / or generate byte code to do so. In any case, it introduces a significant in-direction.

```java
public final class Message implements SdkPojo {
   private static final SdkField MESSAGE = SdkField.builder(STRING).name("Message").getter(Message::message).build();
   private static final List<SdkField> SDK_FIELDS = ImmutableList.of(MESSAGE);

  @Override // from SdkPojo 
  public final List<SdkField<?>> sdkFields() { return SDK_FIELDS; }

  private final String message;

  public final String message() { return message; }
}
```

Reply:
Besides, I'm wondering about the value of the method `name()` in `FieldValueGetter`: It's not used anywhere and prevents using lambdas with `FieldValueGetter`.